Implement struct marshalling via IL Stubs instead of via FieldMarshalers

jkoritzinsky opened the pull request:
Currently, our system for marshalling fields of structures between managed and native code is completely separate from our system for marshalling parameters or return values, even though most of the code in the two systems can be shared. This PR unifies the two systems by removing the field marshalers in favor of using IL stubs and a new `NativeFieldDescriptor` concept which is 9 bytes smaller then the old `FieldMarshaler`s.

Perf numbers:

CoreCLR.dll's size is reduced by ~30kB on Windows x64.

I wrote some microbenchmarks to benchmark marshalling with various types of structs:

<details>
<summary>Struct Definitions</summary>

```csharp
public class Common
{
    public const int NumArrElements = 2;
}
//////////////////////////////struct definition///////////////////////////
[StructLayout(LayoutKind.Sequential)]
public struct InnerSequential
{
    public int f1;
    public float f2;
    public string f3;
}

[StructLayout(LayoutKind.Sequential)]//struct containing one field of array type
public struct InnerArraySequential
{
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = Common.NumArrElements)]
    public InnerSequential[] arr;
}


public struct HFA
{
    public float f1;
    public float f2;
    public float f3;
    public float f4;
}

[StructLayout(LayoutKind.Sequential)]
public unsafe struct FixedBufferClassificationTest
{
    public fixed int arr[3];
    public NonBlittableFloat f;
}

// A non-blittable wrapper for a float value.
// Used to force a type with a float field to be non-blittable
// and take a different code path.
[StructLayout(LayoutKind.Sequential)]
public struct NonBlittableFloat
{
    public NonBlittableFloat(float f)
    {
        arr = new []{f};
    }

    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 1)]
    private float[] arr;
    
    public float F => arr[0];
}

[StructLayout(LayoutKind.Sequential)]
public struct S8
{
    public string name;
    public bool gender;
    [MarshalAs(UnmanagedType.Error)]
    public int i32;
    [MarshalAs(UnmanagedType.Error)]
    public uint ui32;
    [MarshalAs(UnmanagedType.U2)]
    public ushort jobNum;
    [MarshalAs(UnmanagedType.I1)]
    public sbyte mySByte;
}
```
</details>


The following are the perf numbers I got on Windows x64: (CoreRun Master is a local Release build of CoreCLR on commit 402af7bcfe2a167c8cdfbe569ddb73e3def6cb0b.)

``` ini

BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18970
Intel Core i7-7700 CPU 3.60GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=3.0.100-preview9-014004
  [Host]     : .NET Core 3.0.0-preview9-19423-09 (CoreCLR 4.700.19.42102, CoreFX 4.700.19.42104), 64bit RyuJIT
  Job-EFUBVD : .NET Core ? (CoreCLR 5.0.19.42001, CoreFX 4.700.19.35605), 64bit RyuJIT
  Job-WIAUZG : .NET Core ? (CoreCLR 5.0.19.45401, CoreFX 5.0.19.42613), 64bit RyuJIT


```
|                               Method |               Toolchain |         Mean |       Error |      StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------------------------------- |------------------------ |-------------:|------------:|------------:|------:|--------:|------:|------:|------:|----------:|
|                            S8ByValue |          CoreRun Master | 13,962.77 ns | 212.9081 ns | 199.1543 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|                            S8ByValue | CoreRun Struct IL Stubs | 12,518.12 ns | 151.9199 ns | 142.1060 ns |  0.90 |    0.02 |     - |     - |     - |         - |
|                                      |                         |              |             |             |       |         |       |       |       |           |
|          InnerArraySequentialByValue |          CoreRun Master | 27,112.46 ns | 372.1051 ns | 348.0673 ns |  1.00 |    0.00 |     - |     - |     - |         - |
|          InnerArraySequentialByValue | CoreRun Struct IL Stubs | 25,386.27 ns | 503.3039 ns | 446.1658 ns |  0.94 |    0.02 |     - |     - |     - |         - |
|                                      |                         |              |             |             |       |         |       |       |       |           |
| FixedBufferClassificationTestByValue |          CoreRun Master |    144.83 ns |   2.4056 ns |   2.2502 ns |  1.00 |    0.00 |     - |     - |     - |         - |
| FixedBufferClassificationTestByValue | CoreRun Struct IL Stubs |     77.08 ns |   0.7635 ns |   0.6768 ns |  0.53 |    0.01 |     - |     - |     - |         - |


~~The allocation in `InnerArraySequentialByValue` is a `System.RuntimeMethodInfoStub` allocated by the `JIT_GetRuntimeMethodStub` helper for implementing the `ldtoken` instruction that loads the token for the nested `InnerSequential` struct IL stub onto the stack in the `InnerArraySequential` struct IL stub.~~

In addition to normal CoreCLR testing, I've also run the WinForms integration test suite with this local build of CoreCLR to validate that it doesn't break upstream. I've also run the struct marshalling tests with GCStress modes 3 and C.

stephentoub commented:
> The following are the perf numbers I got on Windows x64

What is the perf test actually doing?

> The allocation in InnerArraySequentialByValue is a System.RuntimeMethodInfoStub allocated by the JIT_GetRuntimeMethodStub helper

Why is that an acceptable regression?

jkoritzinsky commented:
The perf test is calling a PInvoke passing the struct by value. I'll post the benchmark on Monday.

Re the allocation: I'm planning on trying to figure out how to cache the allocated object so we only allocate at most once per stub instead of once per stub call. Additionally, the allocation only happens when marshalling an array of non-blittable structures, which in that case already has allocations, so I felt it was still worth putting this up as a draft to get some full CI runs on it before figuring out how to remove the single allocation.

stephentoub commented:
> so I felt it was still worth putting this up as a draft to get some full CI runs on it before figuring out how to remove the single allocation.

Ok. Thanks.

jkoritzinsky commented:
I've updated the perf benchmarks with the current result. I've removed the allocation from each iteration by caching it in the `MethodDesc`'s `LoaderAllocator` (which has the same maximum lifetime as the allocated object).

jkoritzinsky commented:
Perf numbers updated for commit 034db68 where I've changed the `ldtoken StructStub` to `ldftn StructStub` and removed the `RuntimeMethodInfoStub` caching.

jkoritzinsky commented:
@AaronRobinsonMSFT can you make a review pass when you have a chance?

AaronRobinsonMSFT commented on src/vm/class.cpp:
```diff
@@ -1472,12 +1472,13 @@ CorElementType EEClassLayoutInfo::GetNativeHFATypeRaw()
-    const FieldMarshaler *pFieldMarshaler = GetFieldMarshalers();
-    while (numReferenceFields--)
+    const NativeFieldDescriptor *pNativeFieldDescriptors = GetNativeFieldDescriptors();
+    for(UINT i = 0; i < numReferenceFields; ++i)
```
Instead of this, I would use an iterator style of loop.

```c++
    const NativeFieldDescriptor *curr = GetNativeFieldDescriptors();
    const NativeFieldDescriptor *end = curr + numReferenceFields;
    for (; curr != end; ++curr)
    {
       ...
    }
```

AaronRobinsonMSFT commented on src/vm/class.cpp:
```diff
@@ -2925,16 +2924,14 @@ void EEClass::Fixup(DataImage *image, MethodTable *pMT)
-        FieldMarshaler *pFMEnd = (FieldMarshaler*) ((BYTE *)pFM + pInfo->m_numCTMFields*MAXFIELDMARSHALERSIZE);
-        while (pFM < pFMEnd)
+        NativeFieldDescriptor *pFM = pInfo->GetNativeFieldDescriptors();
+        for(UINT i = 0; i < pInfo->m_numCTMFields; ++i)
```
Iterator style here as well.

AaronRobinsonMSFT commented on src/vm/class.cpp:
```diff
@@ -3757,15 +3754,15 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
+            NativeFieldDescriptor *pParentCTMFieldSrcArray = pParentLayoutInfo->GetNativeFieldDescriptors();
+            NativeFieldDescriptor *pParentCTMFieldDestArray = pEEClassLayoutInfoOut->GetNativeFieldDescriptors() + numChildCTMFields;
 
             for (UINT parentCTMFieldIndex = 0; parentCTMFieldIndex < pParentLayoutInfo->m_numCTMFields; parentCTMFieldIndex++)
```
This is basically a call to `memmove()` and would probably be much more efficient. If you can guarentee there is no overlap feel free to use `memcpy()`.

jkoritzinsky replied:
We need to call the copy constructors here to ensure we have the correct fixups for the FieldDesc (and MethodTable) in the NativeFieldDescriptor. In the CoreCLR build they're just `PlainPointer`s so it would be fine to just do a memcpy, but in crossgen they're relative fixups so we need to do the copy.

AaronRobinsonMSFT replied:
Boo. Okay.

AaronRobinsonMSFT commented on src/vm/class.cpp:
```diff
@@ -3883,6 +3879,75 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
+        // This type may qualify for ManagedSequential. Collect managed size and alignment info.
+        if (CorTypeInfo::IsPrimitiveType(corElemType))
+        {
+            pManagedPlacementInfo->m_size = ((UINT32)CorTypeInfo::Size(corElemType)); // Safe cast - no primitive type is larger than 4gb!
```
Can you move comment above the statement?

AaronRobinsonMSFT commented on src/vm/class.cpp:
```diff
@@ -3883,6 +3879,75 @@ VOID EEClassLayoutInfo::CollectLayoutFieldMetadataThrowing(
+    // managed-sequential.
+    // This function also fills in the pManagedPlacementInfo structure for this field.
+    //=======================================================================
+    BOOL CheckIfDisqualifiedFromManagedSequential(CorElementType corElemType, MetaSig &fsig, RawFieldPlacementInfo * pManagedPlacementInfo)
```
~~`CheckIfDisqualifiedFromManagedSequential`~~ -> `IsDisqualifiedFromManagedSequential`

AaronRobinsonMSFT commented on src/vm/dispatchinfo.cpp:
```diff
@@ -749,11 +750,7 @@ void DispatchMemberInfo::SetUpParamMarshalerInfo()
-            FieldDesc* pFld = (FieldDesc*) getFieldHandle.Call_RetLPVOID(&arg);
-            if (pFld)
-                SetUpFieldMarshalerInfo(pFld);
+            // We don't support non-default marshalling behavior for field getter/setter stubs invoked via IDispatch.
```
Perhaps assert this case or is this supposed to be a nop?

jkoritzinsky replied:
This has been a no-op with a TODO since moving to GitHub.

AaronRobinsonMSFT commented on src/vm/dispatchinfo.cpp:
```diff
@@ -2344,8 +2335,18 @@ void DispatchInfo::MarshalParamManagedToNativeRef(DispatchMemberInfo *pMemberInf
+
         // Convert the contents of the managed array into the original SAFEARRAY.
-        OleVariant::MarshalSafeArrayForArrayRef((BASEARRAYREF *)pSrcObj, *V_ARRAYREF(pRefVar), ElementVt,  pElementMT);
+        OleVariant::MarshalSafeArrayForArrayRef((BASEARRAYREF *)pSrcObj, *V_ARRAYREF(pRefVar), ElementVt, pElementMT, pStructMarshalStub != nullptr ? pStructMarshalStub->GetMultiCallableAddrOfCode() : NULL);
```
This is kind of hidden. Pull out the statement into a new line.

AaronRobinsonMSFT replied:
Choose `NULL` or `nullptr` in any one statement, try not to mix them.

AaronRobinsonMSFT commented on src/vm/dispparammarshaler.cpp:
```diff
@@ -251,11 +251,19 @@ void DispParamArrayMarshaler::MarshalNativeToManaged(VARIANT *pSrcVar, OBJECTREF
 
         // Convert the contents of the SAFEARRAY.
-        OleVariant::MarshalArrayRefForSafeArray(pSafeArray, (BASEARRAYREF*)pDestObj, vt, pElemMT);
+        OleVariant::MarshalArrayRefForSafeArray(pSafeArray, (BASEARRAYREF*)pDestObj, vt, pStructMarshalStub != nullptr ? pStructMarshalStub->GetMultiCallableAddrOfCode() : NULL, pElemMT);
```
Extract ternary operator into new statement.

AaronRobinsonMSFT commented on src/vm/dispparammarshaler.cpp:
```diff
@@ -251,11 +251,19 @@ void DispParamArrayMarshaler::MarshalNativeToManaged(VARIANT *pSrcVar, OBJECTREF
+        MethodDesc* pStructMarshalStub = nullptr;
+        if ((vt == VT_RECORD || vt == VTHACK_NONBLITTABLERECORD) && (pElemMT != nullptr && !pElemMT->IsBlittable()))
+        {
+            GCX_PREEMP();
```
Since we are entering Preemptive mode, can't `pDestObj` be moved or is that an out? If this is an out then `*pDestObj` should be `nullptr` and I would like to see that asserted in the contract.

AaronRobinsonMSFT commented on src/vm/dispparammarshaler.cpp:
```diff
@@ -397,7 +415,15 @@ void DispParamRecordMarshaler::MarshalNativeToManaged(VARIANT *pSrcVar, OBJECTRE
             BoxedValueClass = m_pRecordMT->Allocate();
-            FmtClassUpdateCLR(&BoxedValueClass, (BYTE*)pvRecord);
+
+            MethodDesc* pStructMarshalStub;
```
Same comment here about `pDestObj` and the state of it.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -139,11 +142,12 @@ StubSigDesc::StubSigDesc(MethodDesc *pMD, Signature sig, Module *pModule)
         PRECONDITION(pModule != NULL);
     }
-    CONTRACTL_END
+        CONTRACTL_END
```
Realign.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -967,24 +1043,22 @@ class ILStubState : public StubState
 
-#ifdef FEATURE_COMINTEROP
-        if (hasTryCatchForHRESULT)
+        ILStubEHClause tryCatchClause = { 0 };
```
The `0` isn't needed.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -1347,6 +1425,132 @@ class ILStubState : public StubState
+            FALSE,
+            FALSE,
+            dwStubFlags,
+            -1,
```
Can you comment on what the `-1` means here.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -4139,10 +4363,150 @@ static void CreateNDirectStubWorker(StubState*         pss,
+    }
+    CONTRACTL_END;
+
+
```
Remove extra line and below.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -4139,10 +4363,150 @@ static void CreateNDirectStubWorker(StubState*         pss,
+    NativeFieldDescriptor* pFieldDescriptors = pMT->GetLayoutInfo()->GetNativeFieldDescriptors();
+
+
+    for (int i = 0; i < numFields; ++i)
```
Iterator style.

jkoritzinsky replied:
We use the value of `i` a few times below, so I think this loop is better left as an indexing loop.

AaronRobinsonMSFT commented on src/vm/dllimport.cpp:
```diff
@@ -5287,6 +5700,119 @@ MethodDesc* NDirect::CreateFieldAccessILStub(
+    // void (ref byte managedData, byte* nativeData, int marshalAction, ref CleanupWorkListElement cwl)
+    int numParamTokens = 1;
+
+    mdParamDef* pParamTokenArray = (mdParamDef*)_alloca(numParamTokens * sizeof(mdParamDef));
```
Why `_alloca`? The size is a compile time constant.

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -443,23 +511,17 @@ internal static void ClearNative(IntPtr pNative)
-            Debug.Fail("NYI");
-            return IntPtr.Zero;
-        }
+            ReadOnlySpan<char> managed = strManaged;
```
Yes it does. It creates an empty span and this becomes just setting the null terminator.

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -121,6 +121,74 @@ internal static void ClearNative(IntPtr pNative)
+            Span<sbyte> tempBuf = stackalloc sbyte[allocSize];
+
+            originalBuffer.CopyTo(tempBuf);
+            tempBuf[length - 1] = 0;
```
I'm matching the old behavior: https://github.com/dotnet/coreclr/blob/99839c5182a5b28348824d90d4d838e9163b3401/src/vm/fieldmarshaler.cpp#L2645-L2647

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -121,6 +121,74 @@ internal static void ClearNative(IntPtr pNative)
+                cbWritten--;
+            }
+
+            buffer[cbWritten] = 0;
```
We should be writing a full terminator. I'm matching the behavior from beforehand: https://github.com/dotnet/coreclr/blob/99839c5182a5b28348824d90d4d838e9163b3401/src/vm/fieldmarshaler.cpp#L2603

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -121,6 +121,74 @@ internal static void ClearNative(IntPtr pNative)
+            int numChars = strManaged.Length;
+            if (numChars >= length)
+            {
+                numChars = length - 1;
```
I don't believe we have a doc reference for the auto-null-terminator logic.

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -573,26 +635,45 @@ internal static class ObjectMarshaler
-        internal static extern void ConvertToNative(IntPtr dst, IntPtr src, IntPtr pMT, ref CleanupWorkListElement pCleanupWorkList);
+        internal static unsafe IntPtr ConvertSafeHandleToNative(SafeHandle? handle, bool hasCleanupWorkList, ref CleanupWorkListElement? cleanupWorkList)
+        {
+            if (!hasCleanupWorkList)
```
We don't have the `ldstr` instruction implemented in the native IL stub emit infrastructure. I felt it would just be easier to throw from within C# than to implement the instruction along with managing the string lifetimes.

jkoritzinsky commented on src/System.Private.CoreLib/src/System/StubHelpers.cs:
```diff
@@ -1460,6 +1563,17 @@ internal struct NativeVariant
     }  // struct NativeVariant
 
+    // This NativeDecimal type is very similar to the System.Decimal type, except it requires an 8-byte alignment
+    // like the native DECIMAL type instead of the 4-byte requirement of the System.Decimal type.
```
This requirement isn't really documented anywhere. It's just a fall-out of the various implementations. The native implementation has an 8-byte field overlapping the last two 4-byte integers, but (because of the old bad support for explicit layout) the managed implementation of Decimal has only 4-byte integer fields.

AaronRobinsonMSFT commented on src/vm/ilstubresolver.cpp:
```diff
@@ -318,42 +337,67 @@ ILStubResolver::AllocGeneratedIL(
-    NewArrayHolder<CompileTimeState> pNewCompileTimeState    = NULL;
+    if (!UseLoaderHeap())
+    {
+        NewArrayHolder<BYTE>             pNewILCodeBuffer = NULL;
```
Can you pull these command variables outside the branch. I don't see why they need to be different.

jkoritzinsky replied:
The reason the branching here is weird is because we want to be able to use the holders for cleanup in the exceptional/failure case. However, when we don't need to use the loader heap, we allocate via "new" and delete in the `ClearCompileTimeState` method. If we use the loader heap, we want to allocate from the loader heap and use the appropriate holder and only delete when the loader is unloaded (not in `ClearCompileTimeState`). If we don't keep the data after `ClearCompileTimeState`, the JIT will assert when trying to JIT a P/Invoke or struct marshal stub that calls this struct marshal stub.

It's a really awkward case where RAII doesn't quite do what we want unless we create a "multi-holder" type from which we could pick which holder from a set we want to use.

AaronRobinsonMSFT commented:
@jkoritzinsky Can you squash these commits and rebase? Some of the tooling I have only supports 64 commits for review :-/

jkoritzinsky commented:
I've rebased this on master. There were a lot of conflicts from a few months ago that I had to re-resolve as part of that, so a mistake may have snuck in that I didn't catch locally.

AaronRobinsonMSFT commented on tests/src/Interop/LayoutClass/LayoutClassTest.cs:
```diff
@@ -210,6 +225,24 @@ public static bool NestedLayoutClass()
 
+        public static bool RecursiveNativeLayout()
+        {
+            
```
Remove line.

AaronRobinsonMSFT commented on tests/src/Interop/LayoutClass/LayoutClassTest.cs:
```diff
@@ -210,6 +225,24 @@ public static bool NestedLayoutClass()
+            }
+            catch (TypeLoadException)
+            {
+                
```
Remove line.

AaronRobinsonMSFT commented on tests/src/Interop/LayoutClass/LayoutClassNative.cpp:
```diff
@@ -72,3 +72,7 @@ extern "C" DLL_EXPORT BOOL STDMETHODCALLTYPE SimpleNestedLayoutClassByValue(Nest
     return SimpleSeqLayoutClassByRef(&v.str);
 }
+
+extern "C" DLL_EXPORT void Invalid(...)
```
>void [](start = 22, length = 4)

Please explicitly define the calling convention.

AaronRobinsonMSFT commented on src/vm/dispatchinfo.cpp:
```diff
@@ -2344,8 +2335,22 @@ void DispatchInfo::MarshalParamManagedToNativeRef(DispatchMemberInfo *pMemberInf
 
+        PCODE pStructMarshalStubAddress = NULL;
+        GCPROTECT_BEGIN(*pSrcObj);
+        if ((ElementVt == VT_RECORD || ElementVt == VTHACK_NONBLITTABLERECORD) && (pElementMT != nullptr && !pElementMT->IsBlittable()))
```
>(ElementVt == VT_RECORD || ElementVt == VTHACK_NONBLITTABLERECORD) [](start = 12, length = 66)

Shouldn't this imply the latter part of this predicate? Why would they be different?

AaronRobinsonMSFT commented on src/vm/dispatchinfo.cpp:
```diff
@@ -2344,8 +2335,22 @@ void DispatchInfo::MarshalParamManagedToNativeRef(DispatchMemberInfo *pMemberInf
+            GCX_PREEMP();
+
+            MethodDesc* pStructMarshalStub = NDirect::CreateStructMarshalILStub(pElementMT);
+            if (pStructMarshalStub != nullptr)
```
>if (pStructMarshalStub != nullptr) [](start = 12, length = 34)

Why is this okay to be `nullptr`? Can you comment on why this is okay.

jkoritzinsky replied:
I should change that to an assert. It shouldn’t ever be null.

AaronRobinsonMSFT commented on src/vm/olevariant.cpp:
```diff
@@ -1301,7 +1304,7 @@ void OleVariant::MarshalBoolVariantOleToCom(VARIANT *pOleVariant,
 
 void OleVariant::MarshalBoolArrayOleToCom(void *oleArray, BASEARRAYREF *pComArray,
-                                          MethodTable *pInterfaceMT)
+                                          MethodTable *pInterfaceMT, PCODE pManagedMarshalerCode)
```
>pManagedMarshalerCode [](start = 75, length = 21)

It seems very few of these functions are using this parameter. What is the reason for them?

jkoritzinsky replied:
These functions are passed up to the calling code through the `Marshaler` struct in `olevariant.h`. The design there is to have each field be a function pointer that points to each needed marshalling method. As a result, they all need to have the same signature, even if one of the parameters (such as this one) is unused. It's the same reason that every marshaler in olevariant.cpp has a `MethodTable*` parameter when only the interface marshaler uses it.

We can't resolve the struct marshalling stub from the `MethodTable*` in the marshaler since the lookup is slow enough to slow down non-blittable structure array marshalling by at least 600%.

AaronRobinsonMSFT commented on src/vm/mtypes.h:
```diff
@@ -113,4 +113,10 @@ DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_KEYVALUEPAIR,                    KeyValuePair
 DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_EXCEPTION,                       HResultExceptionMarshaler,     true)  // For WinRT, marshal exceptions as Windows.Foundation.HResult
 #endif // FEATURE_COMINTEROP
 
+DEFINE_MARSHALER_TYPE(MARSHAL_TYPE_FIXED_ARRAY, FixedArrayMarshaler, false)
```
>DEFINE_MARSHALER_TYPE [](start = 0, length = 21)

Please align these like the rest of the file.

AaronRobinsonMSFT commented:
>     MarshalInfo(Module* pModule,

Please create a new issue to reduce this argument list. It is close to impossible to understand at instantiation time what all the bools and constants mean. We have `&&` and can now have high efficiency struct passing that can describe what constants are meant to define.

---
Refers to: src/vm/mlinfo.h:456 in 59bb717. [](commit_id = 59bb7176f87ee7bc1b360f2401d92b51715f379b, deletion_comment = False)

AaronRobinsonMSFT commented:
>     BOOL IsWinRTScenario()

Any chance you could keep the function but have it always return FALSE when `FEATURE_COMINTEROP` isn't defined? Would remove a lot of the various `FEATURE_COMINTEROP` checks in `mlinfo.cpp`.

---
Refers to: src/vm/mlinfo.h:675 in 59bb717. [](commit_id = 59bb7176f87ee7bc1b360f2401d92b51715f379b, deletion_comment = False)

AaronRobinsonMSFT commented on src/vm/mlinfo.cpp:
```diff
@@ -1946,7 +1972,11 @@ MarshalInfo::MarshalInfo(Module* pModule,
+            m_type = MARSHAL_TYPE_GENERIC_8;
+#else
+            m_type = MARSHAL_TYPE_GENERIC_4;
+#endif
```
These are technical identical since `(sizeof(void*)==4)` is a constant. Please put this back.

AaronRobinsonMSFT replied:
Throughout.

---
In reply to: [338220905](https://github.com/dotnet/coreclr/pull/26340#discussion_r338220905) [](ancestors = 338220905)

jkoritzinsky replied:
These aren't identical in the case of crossgening for Linux arm32 from a Linux x64 machine using the cross-bitness build of crossgen. Since this code is used when jitting IL stubs with native structure types, we need to make sure the type here matches type on the target architecture, not the host.

AaronRobinsonMSFT commented on src/vm/mlinfo.cpp:
```diff
@@ -2175,11 +2220,28 @@ MarshalInfo::MarshalInfo(Module* pModule,
                             break;
                         }
 #endif // FEATURE_COMINTEROP
+                        
```
Remove.

AaronRobinsonMSFT commented on src/vm/mlinfo.cpp:
```diff
@@ -2208,14 +2269,28 @@ MarshalInfo::MarshalInfo(Module* pModule,
+                            else if (builder)
+                            {
+
+                                m_resID = IDS_EE_BADMARSHALPARAM_STRINGBUILDER;
```
Remove line.

AaronRobinsonMSFT commented on src/vm/ilmarshalers.h:
```diff
@@ -35,16 +35,33 @@ class ILStubMarshalHome
+    MarshalHomeType     m_homeType;
+    bool                m_hasTypeInfo = false;
+    bool                m_unalignedIndirectStore;
+    LocalDesc          m_locDesc;
```
Move this above `m_homeType`.

jkoritzinsky commented:
> > ```
> > BOOL IsWinRTScenario()
> > ```
> 
> Any chance you could keep the function but have it always return FALSE when `FEATURE_COMINTEROP` isn't defined? Would remove a lot of the various `FEATURE_COMINTEROP` checks in `mlinfo.cpp`.
> 
> Refers to: src/vm/mlinfo.h:675 in [59bb717](https://github.com/dotnet/coreclr/commit/59bb7176f87ee7bc1b360f2401d92b51715f379b). [](commit_id = [59bb717](https://github.com/dotnet/coreclr/commit/59bb7176f87ee7bc1b360f2401d92b51715f379b), deletion_comment = False)

I've tried to do this and it only really allows us to remove the FEATURE_COMINTEROP defines around the error cases in four places. The majority of usages have to stay covered since we don't define the WinRT marshalers when FEATURE_COMINTEROP is turned off. Since most usages still have to be covered with ifdefs, I think we might as well leave it as is.

jkoritzinsky commented:
> > ```
> > MarshalInfo(Module* pModule,
> > ```
> 
> Please create a new issue to reduce this argument list. It is close to impossible to understand at instantiation time what all the bools and constants mean. We have `&&` and can now have high efficiency struct passing that can describe what constants are meant to define.
> 
> Refers to: src/vm/mlinfo.h:456 in [59bb717](https://github.com/dotnet/coreclr/commit/59bb7176f87ee7bc1b360f2401d92b51715f379b). [](commit_id = [59bb717](https://github.com/dotnet/coreclr/commit/59bb7176f87ee7bc1b360f2401d92b51715f379b), deletion_comment = False)

Submitted the issue. Link: https://github.com/dotnet/coreclr/issues/27399

jkoritzinsky commented:
@jkotas do you have any feedback on this PR?

jkoritzinsky commented:
Linux_musl x64 failure is #26057
